base/pkg/libgit2/reference.jl

```
@@ -0,0 +1,154 @@
+end
+
+function fullname(ref::GitReference)
+    isempty(ref) && return ""
+    reftype(ref) == GitConst.REF_OID && return ""
+    rname = ccall((:git_reference_symbolic_target, :libgit2), Ptr{UInt8}, (Ptr{Void},), ref.ptr)
+    rname == C_NULL && return ""
+    return bytestring(rname)
+end
+
+function name(ref::GitReference)
+    isempty(ref) && return ""
+    name_ptr = ccall((:git_reference_name, :libgit2), Ptr{UInt8}, (Ptr{Void},), ref.ptr)
+    name_ptr == C_NULL && return ""
+    return bytestring(name_ptr)
+end
+
+function branch(ref::GitReference)
+    isempty(ref) && return ""
+    str_ptr_ptr = Ref{Ptr{UInt8}}(C_NULL)
+    @check ccall((:git_branch_name, :libgit2), Cint,
+                 (Ref{Ptr{UInt8}}, Ptr{Void},), str_ptr_ptr, ref.ptr)
+    return bytestring(str_ptr_ptr[])
+end
+
+function peel(ref::GitReference, obj_type::Cint)
+    obj_ptr_ptr = Ref{Ptr{Void}}(C_NULL)
+    err = ccall((:git_reference_peel, :libgit2), Cint,
+                (Ptr{Ptr{Void}}, Ptr{Void}, Cint), obj_ptr_ptr, ref.ptr, obj_type)
+    if err == Error.ENOTFOUND
+        return Oid()
+    elseif err != Error.GIT_OK
+        if obj_ptr_ptr[] != C_NULL
+            finalize(GitAnyObject(obj_ptr_ptr[]))
+        end
+        throw(Error.GitError(err))
+    end
+    id = Oid(obj_ptr_ptr[])
+    finalize(GitAnyObject(obj_ptr_ptr[]))
+    return id
+end
+
+function create_reference(repo::GitRepo, obj_oid::Oid, refname::AbstractString = GitConst.HEAD_FILE;
+                          force::Bool=false, msg::AbstractString="")
+    sig = default_signature(repo)
+    ref_ptr_ptr = Ref{Ptr{Void}}(C_NULL)
+    try
+        @check ccall((:git_reference_create, :libgit2), Cint,
+                 (Ptr{Ptr{Void}}, Ptr{Void}, Ptr{UInt8}, Ptr{Oid}, Cint, Ptr{SignatureStruct}, Ptr{UInt8}),
+                  ref_ptr_ptr, repo.ptr, refname, Ref(obj_oid), Cint(force),
+                  sig.ptr, isempty(msg) ? Ptr{UInt8}(C_NULL) : msg)
```

See above regarding `Cstring` here and elsewhere.  (`Ptr{UInt8}` is problematic because it will silently truncate any string that has an embedded NUL char: #10994).